### ISSUE TITLE
Reproducing StackLLaMA with Multi-Adapters

### ISSUE THREAD

**Issue by mnoukhov** [huggingface/trl#471]
Now that we've repro'd the StackLLaMA results in #401, I think it would be useful to repro StackLLaMA with the more compute-efficient Multi-Adapter setup #373 to see if we can reduce compute requirements and also as a sort of integration test for Multi-Adapter

**Comment by mnoukhov** [comment]
The multi-adapter fixes make the code better and everything runs but it doesn't repro StackLLaMA. Below is the regular run in brown and the multi-adapter run in yellow. 

The initial values are very similar, which is a good sign, but the training doesn't proceed as expected. 

![image](https://github.com/lvwerra/trl/assets/3391297/4a79c74f-fffd-4d5c-ab4f-6aff92c861d4)

These issues could be related to the KL estimation and perhaps #423 could help. 
Alternatively, the issue could be the difference in quanitization between reward model and rlhf training. Reward modelling uses a `bf16` base LLaMA model whereas we do RLHF training with an `int8` LLaMA model. Currently, I am retraining a reward model with `int8` to see if that's the issue.  


Also, note that we do see reduced memory usage with Multi-Adapter but sadly memory spikes seem just as high

![image](https://github.com/lvwerra/trl/assets/3391297/21f5f8b8-89c1-4ca0-89ca-

**Comment by mnoukhov** [comment]
Hey @dh2shin, if this isn't specifically related to multi-adapter, can you move the discussion back to #401

**Comment by dh2shin** [comment]
@mnoukhov Hi Michael, how has the int8 experiments gone for the reward model? Are you getting decent accuracy.

**Comment by mnoukhov** [comment]
I was running into issues with `ddp` and `peft` (#480 ) but even after getting around them, I haven't managed to make things work. The main thing I found was that you have to train a new reward model adapter on top of the `int8`.

Even after all this, I'm not managing to repro the original results. There could be a regression in the codebase or it could just be too much instability. I'm still investigating but moving to a smaller scale task in the meantime

**Comment by Nipi64310** [comment]
Hi @mnoukhov ,
Peft's set_adapter resembles a [global mode setting](https://github.com/huggingface/peft/blob/e06d94ddeb6c70913593740618df76908b918d66/src/peft/tuners/lora.py#L821). For example,   Process a calculates to the 10th layer using the lora weight of reward. At this point, process b changes the current model's active_adapter to ppo, which causes the ppo adapter to be used for calculation from the 11th to the last layer. I'm not sure if this will happen if you use accelerate launch (deepspeed).
![image](https://github.com/lvwerra/trl/assets/31469418/2d53ed29-e890-412c-99e5-e741f44606db)

**Comment by mnoukhov** [comment]
That is a possible problem and could be solved with a `self.accelerator.wait_for_everyone()`, but I actually wasn't able to re-reproduce StackLLaMA with the latest codebase (without multiple adapters) so that's really the issue. I'm going to retrain an RM and try again

**Comment by wangzhao88** [comment]
hi, when i used your rl model to test the superglue by https://github.com/EleutherAI/lm-evaluation-harness, the result is follow:
| Task    | Version | Metric | Value  |     | Stderr |
|---------|---------|--------|-------:|-----|-------:|
| boolq     | 1       | acc    | 0.7642 | ±   | 0.0074 |
| cb          | 1        | acc    | 0.5536  | ±  | 0.0670 |
|               |           | f1       | 0.4248 |      |              |
| copa      | 0       | acc    | 0.8800 | ±   | 0.0500 |
| multirc  | 1        | acc    | 0.0084  | ±   | 0.0018 |
| record   | 0       | f1       | 0.9119   | ±   | 0.0032 |
|               |          | em     | 0.9044  | ±    | 0.0032 |
| rte         | 0       | acc    | 0.6282 | ±    | 0.0301 |
| wic        | 0       | acc    | 0.4953 | ±    | 0.0198 |
| wsc       | 0       | acc    | 0.5673 | ±    | 0.0474 |

however, when i train my own rl model by:
'''
accelerate launch --multi_gpu --num_machines 1  --num_processes 8 rl_training.py --model_name=【your se model】 --reward_model_name=【your rm model】 --adafactor=False --tokenizer_name=【your se model】 --save_freq=100 --output_max_length=128 --batch_size=8 --gradient_accumulation_steps=8 --batched_gen=True --ppo_epochs=4 --seed=0 --learning_rate=1.4e-5 --early_stopping=True --output_dir=llama-se-rl-finetune-128-8-8-1.4e-5_adam
'''

after one day, the result of my own model(llama-se-rl-finetune-128-8-8-1.4e-5_adamstep_1300) is:
| Task    | Version | Metric | Value  |     | Stderr |
|---------|---------|--------|-------:|-----|-------:|
| boolq     | 1       | acc    | 0.3783 | ±   | 0.0085 |
| cb          | 1        | acc    | 0.4107  | ±  | 0.0663 |
|               |           | f1       | 0.1941  |      |              |
| copa      | 0       | acc    | 0.5500 | ±   | 0.0500 |
| multirc  | 1        | acc    | 0.0031  | ±   | 0.0018 |
| record   | 0       | f1       | 0.1186   | ±   | 0.0032 |
|               |          | em     | 0.1151   | ±    | 0.0032 |
| rte         | 0       | acc    | 0.5271  | ±    | 0.0301 |
| wic        | 0       | acc    | 0.5000 | ±    | 0.0198 |
| wsc       | 0       | acc    | 0.6346 | ±    | 0.0474 |

is means that the training of model is failed?

**Comment by mnoukhov** [comment]
Still relevant, posting so it isn't marked stale.
